build.sc

```diff
@@ -0,0 +1,94 @@
+// build.sc
+
+import mill._, scalalib._, os._
+import mill.api.Loose.Agg
+
+trait CompositeModule extends ScalaModule { outer =>
+
+  def sources = {
+    T.sources { millSourcePath }
+  }
+
+  def allSourceFiles = T {
+    val submodules =
+      millModuleDirectChildren.map(_.millSourcePath)
+    def isHiddenFile(path: os.Path) = path.last.startsWith(".")
+    def isFromSubmodule(path: os.Path) = {
+      submodules.exists { path.startsWith _ }
+    }
+    for {
+      root <- allSources()
+      if os.exists(root.path)
+      path <- (if (os.isDir(root.path)) os.walk(root.path) else Seq(root.path))
+      if os
+        .isFile(path) && ((path.ext == "scala" || path.ext == "java") && !isHiddenFile(path) && !isFromSubmodule(path))
+    } yield PathRef(path)
+  }
+
+  trait Tests extends super.Tests with CompositeModule {
+    def ivyDepsExtra: Agg[Dep] = Agg()
+    final def ivyDeps = ivyDepsExtra ++ testingLibrary
+    def moduleDepsExtra: Seq[JavaModule] = Seq()
+    final def moduleDeps = (Seq(outer) ++ moduleDepsExtra)
+    def testFrameworks = Seq("org.scalatest.tools.Framework")
+    def testingLibrary = Agg(ivy"org.scalatest::scalatest:3.0.5")
+
+  }
+
+}
+
+object deps {
+  val shapeless =
+    Agg(ivy"com.chuusai::shapeless:2.3.3")
```

Comment:
A bit neater to have `val shapeless = ivy"..."` etc then `Agg(shapeless, ...)`.

Reply:
The reason I'm doing it this way is because it allow us to have composite dependencies. For example:

```scala
object deps {
  val circe = {
    val circeVersion = "0.10.0"

    Agg(
      ivy"io.circe::circe-core:$circeVersion",
      ivy"io.circe::circe-generic:$circeVersion",
      ivy"io.circe:circe-parser:$circeVersion"
    )
  }
}

object Foo extends IOHKModule {
  val ivyDeps = deps.circe
}
```